File: P_12/cpp/paralela/conVolParalela.cpp
//
// Created by fedem on 2020-08-11.
//

#include<iostream>
#include<vector>
#include<thread>


using namespace std;
typedef std::vector<std::vector<int> > matrix;

matrix getMatrixM();

matrix getMatrixExample();

matrix getMatrixEnfoque();

matrix getMatrixRealBordes();

matrix getMatrixDetecBordes();

matrix getMatrixSobel();

matrix getMatrixSharpen();

int showMenuAndChooseOption(matrix &M, const matrix &example, const matrix &enfoque, const matrix &realbordes,
                            const matrix &detecbordes, const matrix &sobel, const matrix &sharpen, int op,
                            matrix &kernel, matrix &conv);


ostream &operator<<(ostream &o, const matrix &v) {
    for (int i = 0; i < v.size(); ++i) {
        for (int j = 0; j < v.size(); ++j) {
            cout << v[i][j] << " ";
        }
        cout << endl;
    }
    o << endl << endl;
    return o;
}

int element(matrix &m, int i, int j) {
    return i < 0 || j < 0 || i > m.size() - 1 || j > m.size() - 1 ? 0 : m[i][j];
}

int validResult(int n) {
    if (n < 0)
        return 0;
    if (n > 255)
        return 255;
    else
        return n;
}

int calc(matrix &m, matrix &c, int x, int y) {
    int suma = 0;
    int fil = 0;
    int col;
    for (int i = x - 1; i <= x + 1; i++) {
        col = 0;
        for (int j = y - 1; j <= y + 1; j++) {
            suma += element(m, i, j) * c[fil][col];
            col++;
        }
        fil++;
    }
    return suma;
}


void convolucion(matrix &M, matrix &C, matrix &res, int ini,
                 int end)
{
    int sizeM = M.size();

    for (int i = ini; i < end; i++) {
        res[i].resize(M.size());
        for (int j = 0; j < sizeM; j++) {
            int aux = calc(M, C, i, j);
            res[i][j] = validResult(aux);

        }
    }
}


void llamada(matrix &M, matrix &C, matrix &convo) {
    int msize = M.size();
    int nhilos = 4;
    vector<thread> hilos(4);
    int ventana = msize / nhilos;
    for (int i = 0; i < nhilos - 1; i++) {
        hilos[i] = thread(convolucion, ref(M), ref(C), ref(convo), i * ventana, (i + 1) * ventana); //POR AQUI VOY
    }

    hilos[nhilos - 1] = thread(convolucion, ref(M), ref(C), ref(convo), (nhilos - 1) * ventana, msize);

    for (int i = 0; i < nhilos; i++)
        hilos[i].join();
}


int main(int argc, char const *argv[]) {
    matrix M = getMatrixM();

    matrix example = getMatrixExample();

    matrix enfoque = getMatrixEnfoque();

    matrix realbordes = getMatrixRealBordes();

    matrix detecbordes = getMatrixDetecBordes();

    matrix sobel = getMatrixSobel();

    matrix sharpen = getMatrixSharpen();

    int op = 1;
    matrix kernel;
    matrix conv = M;
    do {
        op = showMenuAndChooseOption(M, example, enfoque, realbordes, detecbordes, sobel, sharpen, op, kernel, conv);
    } while (op != 0);


    return 0;
}

int showMenuAndChooseOption(matrix &M, const matrix &example, const matrix &enfoque, const matrix &realbordes,
                            const matrix &detecbordes, const matrix &sobel, const matrix &sharpen, int op,
                            matrix &kernel, matrix &conv) {
    cout << "Introduce una opcion: " << endl << endl;
    cout << "0-Salir" << endl;
    cout << "1-Ejemplo" << endl;
    cout << "2-Enfoque" << endl;
    cout << "3-Detectar bordes" << endl;
    cout << "4-Realzar bordes" << endl;
    cout << "5-Sobel" << endl;
    cout << "6-Sharpen" << endl;
    cin >> op;
    switch (op) {
        case 0:
            break;
        case 1:
            cout << "Matriz del ejemplo" << endl;
            kernel = example;
            break;
        case 2:
            cout << "Matriz para enfoque" << endl;
            kernel = enfoque;
            break;
        case 3:
            cout << "Matriz para detectar bordes" << endl;
            kernel = detecbordes;
            break;
        case 4:
            cout << "Matriz para realzar bordes" << endl;
            kernel = realbordes;
            break;
        case 5:
            cout << "Matriz Sobel" << endl;
            kernel = sobel;
            break;
        case 6:
            cout << "Matriz Sharpen" << endl;
            kernel = sharpen;
            break;
    }
    cout << endl;

    if (op != 0) {
        cout << "Matriz Original: " << endl;
        cout << M;
        cout << "Matriz Kernel:" << endl;
        cout << kernel;
        llamada(M, kernel, conv);
        cout << "Matriz Convolucionada:" << endl;
        cout << conv;
        cout << endl << endl;
    }
    return op;
}

matrix getMatrixSharpen() {
    matrix sharpen =
            {
                    {1,  -2, 1},
                    {-2, 5,  -2},
                    {1,  -2, 1}
            };
    return sharpen;
}

matrix getMatrixSobel() {
    matrix sobel =
            {
                    {-1, 0, 1},
                    {-2, 0, 2},
                    {-1, 0, 1}
            };
    return sobel;
}

matrix getMatrixDetecBordes() {
    matrix detecbordes =
            {
                    {0, 1,  0},
                    {1, -4, 1},
                    {0, 1,  0}
            };
    return detecbordes;
}

matrix getMatrixRealBordes() {
    matrix realbordes =
            {
                    {0,  0, 0},
                    {-1, 1, 0},
                    {0,  0, 0}
            };
    return realbordes;
}

matrix getMatrixEnfoque() {
    matrix enfoque =
            {
                    {0,  -1, 0},
                    {-1, 5,  -1},
                    {0,  -1, 0}
            };
    return enfoque;
}

matrix getMatrixExample() {
    matrix example =
            {
                    {-2, -1, 0},
                    {-1, 1,  1},
                    {0,  1,  2}
            };
    return example;
}

matrix getMatrixM() {
    matrix M =
            {
                    {35, 40, 41, 45, 50},
                    {40, 40, 42, 46, 52},
                    {42, 46, 50, 55, 55},
                    {48, 52, 56, 58, 60},
                    {56, 60, 65, 70, 75}
            };
    return M;
}

File: P_12/ej2/conVolParalela.java
package ej2;

import java.util.Scanner;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

/** The type Con vol paralela. */
public class conVolParalela implements Runnable {
  /**
   * Print matrix.
   *
   * @param mat the mat
   */
  public static void printMatrix(int[][] mat) {
    int size = mat.length;
    for (int[] ints : mat) {
      for (int j = 0; j < size; j++) {
        System.out.print(ints[j] + " ");
      }
      System.out.println();
    }
    System.out.println();
  }

  /** The Begin. */
  int begin;

  /** The End. */
  int end;

  /** The Matriz. */
  int[][] matriz;

  /** The Kernel. */
  int[][] kernel;

  /** The Solution. */
  int[][] solution;

  /**
   * Instantiates a new Con vol paralela.
   *
   * @param begin the begin
   * @param end the end
   * @param matriz the matriz
   * @param kernel the kernel
   * @param solution the solution
   */
  conVolParalela(int begin, int end, int[][] matriz, int[][] kernel, int[][] solution) {
    this.begin = begin;
    this.end = end;
    this.matriz = matriz;
    this.kernel = kernel;
    this.solution = solution;
  }

  /**
   * Elemento int.
   *
   * @param matrix the matrix
   * @param i the
   * @param j the j
   * @return the int
   */
  int elemento(int[][] matrix, int i, int j) {
    return i < 0 || j < 0 || i > matrix.length - 1 || j > matrix.length - 1 ? 0 : matrix[i][j];
  }

  /**
   * Resultadovalido int.
   *
   * @param n the n
   * @return the int
   */
  int resultadovalido(int n) {
    if (n < 0) {
      return 0;
    }
    return Math.min(n, 255);
  }

  /**
   * Calculo int.
   *
   * @param m the m
   * @param c the c
   * @param x the x
   * @param y the y
   * @return the int
   */
  int calculo(int[][] m, int[][] c, int x, int y) {
    int suma = 0;
    int fil = 0;
    int col = 0;
    for (int i = x - 1; i <= x + 1; i++) {
      col = 0;
      for (int j = y - 1; j <= y + 1; j++) {
        suma += elemento(m, i, j) * c[fil][col];
        col++;
      }
      fil++;
    }
    return suma;
  }

  public void run() {
    for (int i = begin; i < end; i++) {
      for (int j = 0; j < matriz.length; j++) {
        solution[i][j] = resultadovalido(calculo(matriz, kernel, i, j));
      }
    }
    System.out.println("recibo:" + begin + "-" + end);
  }

  private static int[][] getMatrixDetecBordes() {
    return new int[][] {
      {0, 1, 0},
      {1, -4, 1},
      {0, 1, 0}
    };
  }

  private static int[][] getMatrix(int i2, int i3, int i4, int i5, int i6, int i7) {
    return new int[][] {
      {i2, i3, i4},
      {-i5, i6, i7},
      {i2, i3, i4}
    };
  }

  private static int[][] getMatrixEnfoque() {
    return new int[][] {
      {0, -1, 0},
      {-1, 5, -1},
      {0, -1, 0}
    };
  }

  private static int[][] getMatrixExample() {
    return new int[][] {
      {-2, -1, 0},
      {-1, 1, 1},
      {0, 1, 2}
    };
  }

  private static int[][] getMatrixM() {
    return new int[][] {
      {35, 40, 41, 45, 50},
      {40, 40, 42, 46, 52},
      {42, 46, 50, 55, 55},
      {48, 52, 56, 58, 60},
      {56, 60, 65, 70, 75}
    };
  }

  /**
   * The entry point of application.
   *
   * @param args the input arguments
   */
  public static void main(String[] args) {
    Scanner scanner = new Scanner(System.in);

    int[][] M = getMatrixM();

    int[][] ejemplo = getMatrixExample();

    int[][] enfoque = getMatrixEnfoque();

    int[][] realbordes = getMatrix(0, 0, 0, 1, 1, 0);

    int[][] detecbordes = getMatrixDetecBordes();

    int[][] sobel = getMatrix(-1, 0, 1, 2, 0, 2);

    int[][] sharpen = getMatrix(1, -2, 1, 2, 5, -2);

    int op;
    int[][] kernel = new int[0][0];
    int[][] convolucionada = new int[M.length][M.length];
    do {

      System.out.println("Introduce una opcion: ");
      System.out.println();
      System.out.println();
      System.out.println("0-Salir");
      System.out.println("1-Ejemplo");
      System.out.println("2-Enfoque");
      System.out.println("3-Detectar bordes");
      System.out.println("4-Realzar bordes");
      System.out.println("5-Sobel");
      System.out.println("6-Sharpen");
      op = scanner.nextInt();
      switch (op) {
        case 0:
          break;
        case 1:
          System.out.println("Matriz del ejemplo");
          kernel = ejemplo;
          break;
        case 2:
          System.out.println("Matriz para enfoque");
          kernel = enfoque;
          break;
        case 3:
          System.out.println("Matriz para detectar bordes");
          kernel = detecbordes;
          break;
        case 4:
          System.out.println("Matriz para realzar bordes");
          kernel = realbordes;
          break;
        case 5:
          System.out.println("Matriz Sobel");
          kernel = sobel;
          break;
        case 6:
          System.out.println("Matriz Sharpen");
          kernel = sharpen;
          break;
      }

      System.out.println();
      if (op != 0) {
        System.out.println("Matriz Original");
        printMatrix(M);
        System.out.println("Matriz Kernel");
        printMatrix(kernel);

        int size = M.length;
        int nThreads = Runtime.getRuntime().availableProcessors();
        conVolParalela[] tareas = new conVolParalela[nThreads];
        int window = size / nThreads;
        int nTasks = tareas.length;

        ExecutorService newFixedThreadPool = Executors.newFixedThreadPool(nThreads);

        for (int i = 0; i < nTasks - 1; i++) {
          tareas[i] = new conVolParalela(i * window, (i + 1) * window, M, kernel, convolucionada);
          newFixedThreadPool.execute(tareas[i]);
        }

        tareas[nTasks - 1] =
            new conVolParalela((nTasks - 1) * window, size, M, kernel, convolucionada);

        newFixedThreadPool.execute(tareas[nTasks - 1]);

        newFixedThreadPool.shutdown();

        try {
          newFixedThreadPool.awaitTermination(20000, TimeUnit.MILLISECONDS);
        } catch (InterruptedException I) {
          I.printStackTrace();
        }

        System.out.println("Matriz Convolucionada");
        printMatrix(convolucionada);
        System.out.println();
        System.out.println();
      }

    } while (op != 0);
  }
}
